WalletWasabi/Stores/IndexStore.cs

```diff
@@ -192,204 +121,110 @@ private bool TryProcessFilterNoLock(FilterModel filter, bool enqueue)
 
 	public async Task AddNewFiltersAsync(IEnumerable<FilterModel> filters)
 	{
-		var successAny = false;
-
-		foreach (var filter in filters)
+		if (NewFilter is null)
 		{
-			var success = false;
-
-			using (await IndexLock.LockAsync(CancellationToken.None).ConfigureAwait(false))
-			{
-				success = TryProcessFilterNoLock(filter, enqueue: true);
-			}
+			// Lock once.
+			using IDisposable lockDisposable = await IndexLock.LockAsync(CancellationToken.None).ConfigureAwait(false);
 
-			successAny = successAny || success;
+			using SqliteTransaction sqliteTransaction = IndexStorage.BeginTransaction();
 
-			if (success)
+			foreach (FilterModel filter in filters)
 			{
-				NewFilter?.Invoke(this, filter); // Event always outside the lock.
+				if (!TryProcessFilterNoLock(filter, enqueue: true))
+				{
+					sqliteTransaction.Rollback();
```

Comment:
Not needed because an exception thrown on the next line will return the control to a higher caller and cause `finally` part of `using` being executed. That in turn calls `Dispose` on the transaction which automatically rollbacks ([source](https://github.com/dotnet/efcore/blob/e007d7c390081b943347699434f8ad6c65e9b0a4/src/Microsoft.Data.Sqlite.Core/SqliteTransaction.cs#L202-L210)).